lib/SystemPruneCaches.js

```diff
@@ -0,0 +1,161 @@
+  if (_stat.isFile()) {
+    return _stat.size;
+  }
+
+  if (_stat.isDirectory()) {
+    const names = await readdir(dir);
+    let size = 0;
+    for (const name of names) {
+      size += await directorySize(join(dir, name));
+    }
+    return size;
+  }
+
+  return 0;
+};
+
+class CacheInfo {
+  constructor(id = '') {
+    this.id = id;
+    this.lastModified = 0;
+    this.size = 0;
+  }
+
+  static async fromDirectory(dir) {
+    const info = new CacheInfo(basename(dir));
+    info.lastModified = +(await stat(join(dir, 'stamp'))).mtime;
+    info.size = await directorySize(dir);
+    return info;
+  }
+
+  static async fromDirectoryChildren(dir) {
+    const children = [];
+    const names = await readdir(dir);
+    for (const name of names) {
+      children.push(await CacheInfo.fromDirectory(join(dir, name)));
+    }
+    return children;
+  }
+}
+
+// Compilers for webpack with multiple parallel configurations might try to
+// delete caches at the same time. Mutex lock the process of pruning to keep
+// from multiple pruning runs from colliding with each other.
+let deleteLock = null;
+
+class PruneCachesSystem {
+  constructor(cacheRoot, options = {}) {
+    this.cacheRoot = cacheRoot;
+
+    this.options = Object.assign(
+      {
+        // Caches younger than `maxAge` are not considered for deletion. They
+        // must be at least this (default: 2 days) old in milliseconds.
+        maxAge: 2 * 24 * 60 * 60 * 1000,
+        // All caches together must be larger than `sizeThreshold` before any
+        // caches will be deleted. Together they must be at least this
+        // (default: 50 MB) in big in bytes.
+        sizeThreshold: 50 * 1024 * 1024,
+      },
+      options,
+    );
+  }
+
+  apply(compiler) {
+    const compilerHooks = pluginCompat.hooks(compiler);
+
+    const deleteOldCaches = async () => {
+      while (deleteLock !== null) {
+        await deleteLock;
+      }
+
+      let resolveLock;
+
+      let infos;
+      try {
+        deleteLock = new Promise(resolve => {
+          resolveLock = resolve;
+        });
+
+        infos = await CacheInfo.fromDirectoryChildren(this.cacheRoot);
+
+        // Sort lastModified in descending order. More recently modified at the
+        // beginning of the array.
+        infos.sort((a, b) => b.lastModified - a.lastModified);
+
+        const totalSize = infos.reduce((carry, info) => carry + info.size, 0);
+        const oldInfos = infos.filter(
+          info => info.lastModified < Date.now() - this.options.maxAge,
+        );
+        const oldTotalSize = oldInfos.reduce(
+          (carry, info) => carry + info.size,
+          0,
+        );
+
+        if (oldInfos.length > 0 && totalSize > this.options.sizeThreshold) {
+          const newInfos = infos.filter(
+            info => info.lastModified >= Date.now() - this.options.maxAge,
+          );
+
+          for (const info of oldInfos) {
+            rimraf(join(this.cacheRoot, info.id));
```

Comment:
Not sure the state on rimraf. The build is failing and there's an ignored issue from a month ago with folks wondering what's up with the state of the project. Since you're not really using any of the more "advanced" features of rimraf maybe it's not a dep you need?

Reply:
Interesting. Looks like the build fail is use of const and rimraf 9 months ago was still testing for node `0.10` and `0.12` support which don't support `const`. `hard-source` is supporting back to node `8` so I'm not worried about **old** node incompatibility.

I'm not sure if I want to replace rimraf. `hard-source` has been using the current version since December (2017). https://www.npmjs.com/search?q=keywords%3Arm is a fun search to look for alternatives though.